src/rules/noUnnecessaryClassRule.ts

```diff
@@ -0,0 +1,126 @@
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-unnecessary-class",
+        description: Lint.Utils.dedent`
+            Disallows classes containing only static members. Classes
+            with non-empty constructors are ignored.`,
+        rationale: Lint.Utils.dedent`
+            Users who come from a Java-style OO language may wrap
+            their utility functions in an extra class, instead of
+            putting them at the top level.`,
+        optionsDescription: Lint.Utils.dedent`
+            Three arguments may be optionally provided:
+
+            * \`"allow-constructor-only"\` ignores classes whose members are constructors.
+            * \`"allow-empty-class"\` ignores \`class DemoClass {}\`.
+            * \`"allow-static-only"\` ignores classes whose members are static.`,
+        options: {
+            type: "array",
+            items: {
+                type: "string",
+            },
+            minLength: 0,
+            maxLength: 3,
+        },
+        optionExamples: [true, ["allow-empty-class", "allow-constructor-only"]],
+        type: "functionality",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_CONSTRUCTOR_ONLY = "Every member of this class is a constructor. Use functions instead.";
+    public static FAILURE_STATIC_ONLY = "Every member of this class is static. Use namespaces or plain objects instead.";
+    public static FAILURE_EMPTY_CLASS = "This class has no members.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NoUnnecessaryClassWalker(sourceFile, this.ruleName, this.ruleArguments));
+    }
+}
+
+const OPTION__ALLOW_CONSTRUCTOR_ONLY = "allow-constructor-only";
+const OPTION__ALLOW_EMPTY_CLASS = "allow-empty-class";
+const OPTION__ALLOW_STATIC_ONLY = "allow-static-only";
+
+class NoUnnecessaryClassWalker extends Lint.AbstractWalker<string[]> {
+    public walk(sourceFile: ts.SourceFile) {
+        const checkIfStaticOnlyClass = (node: ts.Node): void => {
+            if (isClassDeclaration(node) && !hasExtendsClause(node)) {
+                if (node.members.length === 0 && !this.hasOption(OPTION__ALLOW_EMPTY_CLASS)) {
+                    this.addFailureAtNode(getChildOfKind(node, ts.SyntaxKind.ClassKeyword)!, Rule.FAILURE_EMPTY_CLASS);
+                    return;
+                } else if (node.members.length == 0) {
+                    return;
+                }
+
+                if (node.members.some(isConstructorWithClassDeclaration)) {
+                    return ts.forEachChild(node, checkIfStaticOnlyClass);
+                }
+
+                const allMembersAreConstructors = node.members.every(isConstructorDeclaration);
+
+                if (allMembersAreConstructors && !this.hasOption(OPTION__ALLOW_CONSTRUCTOR_ONLY)) {
```

Comment:
You need to check if the constructor has parameter properties, either by default or add an option for that.